src/platform/unix/mod.rs

```diff
@@ -435,61 +443,76 @@ impl OsIpcReceiverSet {
         let fnv = BuildHasherDefault::<FnvHasher>::default();
         Ok(OsIpcReceiverSet {
             incrementor: Incrementor::new(),
-            pollfds: Vec::new(),
-            fdids: HashMap::with_hasher(fnv)
+            poll: match Poll::new() {
+                Ok(poll) => poll,
+                Err(e) => { return Err(e.into()); }
+            },
+            pollfds: HashMap::with_hasher(fnv),
+            events: Events::with_capacity(10)
         })
     }
 
     pub fn add(&mut self, receiver: OsIpcReceiver) -> Result<u64,UnixError> {
         let last_index = self.incrementor.increment();
         let fd = receiver.consume_fd();
-        self.pollfds.push(pollfd {
-            fd: fd,
-            events: POLLIN,
-            revents: 0,
-        });
-        self.fdids.insert(fd, last_index);
-        Ok(last_index)
+        let io = EventedFd(&fd);
+        let fd_token = Token(fd as usize);
+        let poll_entry = PollEntry {
+            id: last_index,
+            fd: fd
+        };
+        match self.poll.register(&io,
+                                 fd_token,
+                                 Ready::readable(),
+                                 PollOpt::level()) {
+            Ok(_) => {
+                self.pollfds.insert(fd_token, poll_entry);
+                Ok(last_index)
+            },
+            Err(e) => {
+                Err(e.into())
```

Comment:
This could also use `try!()` I think? Though the win here is slightly less clear than above I guess...